source/libuv/uv_event_loop.c

```diff
@@ -0,0 +1,792 @@
+    }
+}
+
+static void s_uv_close_sub_async(uv_handle_t *handle) {
+
+    struct handle_data *handle_data = handle->data;
+    struct libuv_loop *impl = handle_data->event_loop->impl_data;
+
+    handle->data = impl;
+    s_uv_close_handle(handle);
+}
+
+static void s_uv_async_poll_start(uv_async_t *handle) {
+
+    struct handle_data *handle_data = handle->data;
+    struct libuv_loop *impl = handle_data->event_loop->impl_data;
+
+    uv_poll_init(impl->uv_loop, &handle_data->poll, handle_data->owner->data.fd);
+    uv_poll_start(&handle_data->poll, handle_data->uv_events, s_uv_poll_cb);
+
+    uv_close((uv_handle_t *)handle, s_uv_close_sub_async);
+}
+
+static int s_subscribe_to_io_events(
+    struct aws_event_loop *event_loop,
+    struct aws_io_handle *handle,
+    int events,
+    aws_event_loop_on_event_fn *on_event,
+    void *user_data) {
+
+    assert(event_loop);
+    assert(handle->data.fd != -1);
+    assert(handle->additional_data == NULL);
+    assert(on_event);
+    /* Must subscribe for read, write, or both */
+    assert(events & (AWS_IO_EVENT_TYPE_READABLE | AWS_IO_EVENT_TYPE_WRITABLE));
+
+    struct libuv_loop *impl = event_loop->impl_data;
+
+    aws_atomic_fetch_add(&impl->num_open_handles, 1);
+
+    struct handle_data *handle_data = aws_mem_acquire(event_loop->alloc, sizeof(struct handle_data));
+    if (!handle_data) {
+        return AWS_OP_ERR;
+    }
+
+    AWS_ZERO_STRUCT(*handle_data);
+
+    handle_data->poll.data = handle_data;
+    handle_data->owner = handle;
+    handle_data->event_loop = event_loop;
+    handle_data->on_event = on_event;
+    handle_data->on_event_user_data = user_data;
+
+    aws_linked_list_push_back(&impl->el_thread_data.open_subscriptions, &handle_data->open_subs_node);
+
+    handle->additional_data = handle_data;
+
+    handle_data->uv_events = UV_DISCONNECT;
+    if (events & AWS_IO_EVENT_TYPE_READABLE) {
+        handle_data->uv_events |= UV_READABLE;
+    }
+    if (events & AWS_IO_EVENT_TYPE_WRITABLE) {
+        handle_data->uv_events |= UV_WRITABLE;
+    }
+
+    if (s_is_on_callers_thread(event_loop)) {
+        /* If on UV thread, directly start sub */
+        uv_poll_init(impl->uv_loop, &handle_data->poll, handle_data->owner->data.fd);
+        uv_poll_start(&handle_data->poll, handle_data->uv_events, s_uv_poll_cb);
+    } else {
+        /* Otherwise, schedule async to do sub */
+        aws_atomic_fetch_add(&impl->num_open_handles, 1);
+
+        handle_data->poll_start_async.data = handle_data;
+        uv_async_init(impl->uv_loop, &handle_data->poll_start_async, s_uv_async_poll_start);
+        uv_async_send(&handle_data->poll_start_async);
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_unsubscribe_from_io_events(struct aws_event_loop *event_loop, struct aws_io_handle *handle) {
+
+    (void)event_loop;
+    assert(handle->additional_data);
+
+    struct handle_data *handle_data = handle->additional_data;
+
+    if (uv_poll_stop(&handle_data->poll)) {
```

Comment:
likewise, should we do less in the case that the loop is stopped? or is it fine to stop/close stuff that's not currently open